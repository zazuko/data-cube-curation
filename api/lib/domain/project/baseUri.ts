import urlSlug from 'url-slug'
import { URL } from 'url'

export const temporaryUri = (name: string) => `http://cube.data/${urlSlug(name)}/`

export const ensureSlash = <T = string | undefined>(uri: T): T extends undefined ? string | undefined : string => {
  if (typeof uri === 'string') {
    try {
      return new URL(uri.toString()).toString() as any
    } catch (e) {}
  }

  return uri as any
}
