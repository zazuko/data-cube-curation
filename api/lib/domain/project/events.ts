export interface ProjectEvents {
  ProjectCreated: {
    name: string;
    baseUri: string;
  };
  ProjectRenamed: {
    name: string;
  };
  ProjectRebased: {
    baseUri: string;
  };
  ProjectArchived: null;
  FactTableSourceSelected: {
    sourceId: string;
    tableName: string;
  };
  FactTableUnselected: {
    previousSourceId: string;
  };
}
