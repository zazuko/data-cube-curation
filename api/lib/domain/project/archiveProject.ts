import { mutate } from '@tpluscode/fun-ddr'
import { ProjectEvents } from './events'
import { Project } from './index'

export const archiveProject = mutate<Project, never, ProjectEvents>(function (state, cmd, emitter) {
  if (state.archived === 'false') {
    emitter.emit.ProjectArchived(null)
  }

  return {
    ...state,
    archived: 'true',
  }
})
