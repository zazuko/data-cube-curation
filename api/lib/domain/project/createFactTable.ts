import { mutate } from '@tpluscode/fun-ddr'
import { Project } from './index'
import { ProjectEvents } from './events'
import { sources } from '../../storage/repository'
import { errorFactory } from '../error-helper'

interface CreateFactTableCommand {
  sourceId: string
  tableName: string
}

export const selectFactTableSource = mutate<Project, CreateFactTableCommand>(async (project, command, emitter) => {
  const DomainError = errorFactory(project, 'Cannot initialise fact table')

  if (project.factTableSource) {
    throw new DomainError('Fact table has already been selected.')
  }

  const source = await sources.load(command.sourceId).then(ar => ar.state)
  if (!source) {
    throw new DomainError(`Source '${command.sourceId}' was not found`)
  }
  if (source.project !== project['@id']) {
    throw new DomainError(`Source '${command.sourceId}' does not belong to project '${project['@id']}'`)
  }

  emitter.emit<ProjectEvents, 'FactTableSourceSelected'>('FactTableSourceSelected', {
    sourceId: command.sourceId,
    tableName: command.tableName,
  })

  return {
    ...project,
    factTableSource: command.sourceId,
  }
})
