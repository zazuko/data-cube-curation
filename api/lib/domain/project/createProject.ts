import uuid from 'uuid'
import { DomainError, initialize } from '@tpluscode/fun-ddr'
import { ProjectEvents } from './events'
import { Project } from './index'
import { ensureSlash, temporaryUri } from './baseUri'

interface CreateCommand {
  uriSlug?: string;
  name: string;
  baseUri?: string;
}

export const createProject = initialize<Project, CreateCommand, ProjectEvents>(function (createCommand, emitter) {
  if (!createCommand.name || typeof createCommand.name !== 'string') {
    throw new DomainError('', 'Cannot create Project', 'Invalid name')
  }

  const baseUri = ensureSlash(createCommand.baseUri) || temporaryUri(createCommand.name)
  emitter.emit.ProjectCreated({
    name: createCommand.name,
    baseUri,
  })

  return {
    '@id': `/project/${createCommand.uriSlug || uuid()}`,
    '@type': 'Project',
    name: createCommand.name,
    archived: 'false',
    baseUri,
  }
})
