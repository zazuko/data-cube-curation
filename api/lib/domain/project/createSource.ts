import { factory } from '@tpluscode/fun-ddr'
import { Source } from '../source'
import { SourceEvents } from '../source/events'
import { Project } from './index'

interface UploadSourceCommand {
  fileName: string;
  type: 'csv' | 'excel';
  columns: string[];
  sample: string[][];
}

export const createSource = factory<Project, UploadSourceCommand, Source>(function (project, command, emitter) {
  const sourceId = `${project['@id']}/source/${command.fileName}`

  emitter.emit<SourceEvents, 'SourceUploaded'>('SourceUploaded', {
    fileName: command.fileName,
    projectId: project['@id'],
    columns: command.columns,
    sampleRows: command.sample,
  })

  return {
    '@id': sourceId,
    '@type': [ 'Source', 'CsvSource' ],
    type: command.type,
    name: command.fileName,
    project: project['@id'],
    columns: command.columns,
  }
})
