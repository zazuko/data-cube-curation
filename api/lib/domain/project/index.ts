import uuid from 'uuid'
import { mutate, bootstrap, factory, Entity } from '../../ddd'
import { ProjectEvents } from './events'

export interface Project extends Entity {
  name: string;
  archived: boolean;
}

interface Column extends Entity {
  name: string;
}

export interface Source extends Entity {
  type: 'csv' | 'excel';
  project: string;
  columns: Column[];
}

interface CreateCommand {
  uriSlug?: string;
  name: string;
}

export const createProject = bootstrap<Project, CreateCommand>(function (createCommand, emitter) {
  emitter.emit<ProjectEvents, 'ProjectCreated'>('ProjectCreated', {
    name: createCommand.name,
  })

  return {
    '@id': `/project/${createCommand.uriSlug || uuid()}`,
    '@type': 'Project',
    name: createCommand.name,
    archived: false,
  }
})

interface RenameCommand {
  newName: string;
}

export const renameProject = mutate<Project, RenameCommand>(function (state, renameCommand, emitter) {
  if (state.name !== renameCommand.newName) {
    emitter.emit<ProjectEvents, 'ProjectRenamed'>('ProjectRenamed', {
      name: renameCommand.newName,
    })
  }

  return {
    ...state,
    name: renameCommand.newName,
  }
})

interface UploadSourceCommand {
  fileName: string;
  type: 'csv' | 'excel';
  columns: string[];
}

export const createSource = factory<Project, UploadSourceCommand, Source>(function (project, command) {
  const sourceId = `${project['@id']}/source/${command.fileName}`

  return {
    '@id': sourceId,
    '@type': [ 'Source', 'CsvSource' ],
    type: command.type,
    name: command.fileName,
    project: project['@id'],
    columns: command.columns.map(name => ({
      '@id': `${sourceId}/${name}`,
      '@type': 'Column',
      name,
    })),
  }
})

export const archiveProject = mutate<Project, never>(function (state, cmd, emitter) {
  if (!state.archived) {
    emitter.emit<ProjectEvents, 'ProjectArchived'>('ProjectArchived', {})
  }

  return {
    ...state,
    archived: true,
  }
})
