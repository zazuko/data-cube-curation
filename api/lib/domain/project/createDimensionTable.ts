import { factory } from '@tpluscode/fun-ddr'
import { Project } from './index'
import { TableCommand } from './createFactTable'
import { errorFactory } from '../error-helper'
import { DimensionTable, Table } from '../table'
import { TableEvents } from '../table/events'
import { hasSource } from '../../read-graphs/project'
import { extractColumnIds } from '../table/identifierTemplate'

interface CreateDimensionTableCommand extends TableCommand {
  identifierTemplate: string;
}

export const addDimensionTable = factory<Project, CreateDimensionTableCommand, Table>(async (project, cmd, emitter) => {
  const DomainError = errorFactory(project, 'Cannot create dimension table')
  if (!cmd.tableName) {
    throw new DomainError('Name missing')
  }
  if (!cmd.identifierTemplate) {
    throw new DomainError('Identifier template missing')
  }
  if (!await hasSource(project['@id'], cmd.sourceId)) {
    throw new DomainError('Source does not belong to the project')
  }

  const columnIds = await extractColumnIds(cmd.sourceId, cmd.identifierTemplate)
  if (columnIds instanceof Error) {
    throw new DomainError(`Template is not correct. ${columnIds.message}`)
  }

  const tableId = `${project['@id']}/table/${encodeURIComponent(cmd.tableName)}`

  const table: DimensionTable = {
    '@id': tableId,
    projectId: project['@id'],
    '@type': 'Table',
    sourceId: cmd.sourceId,
    tableName: cmd.tableName,
    identifierTemplate: cmd.identifierTemplate,
  }

  emitter.emit<TableEvents, 'DimensionTableCreated'>('DimensionTableCreated', table)

  return table
})
