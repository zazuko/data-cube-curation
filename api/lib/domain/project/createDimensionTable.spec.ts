/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
import { addDimensionTable } from './createDimensionTable'
import { Project } from './index'
import { DomainError } from '@tpluscode/fun-ddr'
import { hasSource } from '../../read-graphs/project'
import { extractColumnIds } from '../table/identifierTemplate'

jest.mock('../../read-graphs/project')
jest.mock('../table/identifierTemplate')
const projectHasSourceMock = hasSource as jest.Mock
const extractColumnIdsMock = extractColumnIds as jest.Mock

describe('project', () => {
  beforeAll(() => {
    projectHasSourceMock.mockResolvedValue(true)
    extractColumnIdsMock.mockResolvedValue([])
  })

  describe('create dimension table', () => {
    it('creates a dataCube:Table aggregate', async () => {
      // given
      const project = {
        '@id': '/project/test',
      } as Project

      // when
      const result = await addDimensionTable(project, {
        sourceId: '/project/test/source',
        tableName: 'Table One',
        identifierTemplate: '/dimension/{column}',
      })

      // then
      await expect(result.state).resolves.toMatchSnapshot()
    })

    it('publishes an event', async () => {
      // given
      const project = {
        '@id': '/project/test',
      } as Project

      // when
      const result = await addDimensionTable(project, {
        sourceId: '/project/test/source',
        tableName: 'Table One',
        identifierTemplate: '/dimension/{column}',
      })

      // then
      await expect(result.events).resolves.toMatchSnapshot()
    })

    it('throws when name is missing', async () => {
      // given
      const project = {
        '@id': '/project/test',
      } as Project
      const command = {
        sourceId: '/project/test/source',
        identifierTemplate: '/dimension/{column}',
      } as any

      // when
      const result = await addDimensionTable(project, command)

      // then
      await expect(result.error).resolves.toBeInstanceOf(DomainError)
    })

    it('throws when source does not belong to the project', async () => {
      // given
      const project = {
        '@id': '/project/test',
      } as Project
      const command = {
        sourceId: '/project/test/source',
        tableName: 'Test',
        identifierTemplate: '/dimension/{column}',
      }
      projectHasSourceMock.mockResolvedValueOnce(false)

      // when
      const result = await addDimensionTable(project, command)

      // then
      await expect(result.error).resolves.toBeInstanceOf(DomainError)
    })

    it('throws when template did not pass validation', async () => {
      // given
      const project = {
        '@id': '/project/test',
      } as Project
      const command = {
        sourceId: '/project/test/source',
        tableName: 'Test',
        identifierTemplate: '/dimension/{column}',
      }
      extractColumnIdsMock.mockResolvedValueOnce(new Error())

      // when
      const result = await addDimensionTable(project, command)

      // then
      await expect(result.error).resolves.toBeInstanceOf(DomainError)
    })
  })
})
