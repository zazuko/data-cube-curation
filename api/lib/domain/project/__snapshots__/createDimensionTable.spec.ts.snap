// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project create dimension table creates a dataCube:Table aggregate 1`] = `
Object {
  "@id": "/project/test/table/Table%20One",
  "@type": "Table",
  "identifierTemplate": "/dimension/{column}",
  "projectId": "/project/test",
  "sourceId": "/project/test/source",
}
`;

exports[`project create dimension table publishes an event 1`] = `
Array [
  Object {
    "data": Object {
      "@id": "/project/test/table/Table%20One",
      "@type": "Table",
      "identifierTemplate": "/dimension/{column}",
      "projectId": "/project/test",
      "sourceId": "/project/test/source",
    },
    "name": "DimensionTableCreated",
  },
]
`;
